### PR TITLE
LPS-120545 Fix ckeditor_classic

### PR DESCRIPTION
**Pull request by julien**
This has a couple of fixes for our `ClassicEditor.js` component:

Avoid [`CKEDITOR.on('instanceCreated', ...)`](https://github.com/liferay-frontend/liferay-portal/pull/390/files#diff-60128641efe40efd4a4df0325f6e6070L118) because this will be invoked for each (created) CKEDITOR instance - which means that before this fix all editors had the same `name` property and the `CKEDITOR.instance` object will only ever contain one key.

Prefer passing props that the component supports like `onDrop` and `onInstanceCreated`

Add a temporary `textarea` element to `ckeditor_classic.jsp` because at the moment, the `CKEditor` (from `ckeditor4-react`) does not set an `id` or `name` to the element it renders, and this is a problem with the [InputLocalized](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/input_localized.js) component that tries to use that element.

**Before this change, this would cause errors**:

**After this change, it looks like this**:

**Test plan**:

- Add : 
  ```properties
  editor.wysiwyg.portal-impl.portlet.ddm.text_html.ftl=ckeditor_classic
  editor.wysiwyg.portal-web.docroot.html.taglib.ui.email_notification_settings.jsp=ckeditor_classic
  ``` 
  to your `portal-ext.properties` file
- Fetch this branch
- Deploy the `frontend-editor-ckeditor-web` module
- Navigate to Web Content, click on the **+** button in the top right and choose Basic Web Content
- Assert that no errors are present when typing text in any of the editors present in the page
- From the global menu, select the Control Panel tab, click on Instance Settings and choose Email
- Assert that no errors are present when typing text in any of the editors present in the page

### CODE REVIEW

**Comment by carloslancha** [comment]
To me the idea for `ckeditor_classic` is not to reproduce everything present in current `ckeditor` but to add things that are needed in the best way even if they don't follow the same API and we need to change the usages.

File: modules/apps/journal/journal-web/src/main/resources/META-INF/resources/article/basic_info.jsp [hunk 1]
```diff
@@ -93,7 +93,7 @@ DDMStructure ddmStructure = journalEditArticleDisplayContext.getDDMStructure();
 		availableLocales="<%= journalEditArticleDisplayContext.getAvailableLocales() %>"
 		cssClass="form-control"
 		defaultLanguageId="<%= journalEditArticleDisplayContext.getDefaultArticleLanguageId() %>"
-		editorName="ckeditor"
+		editorName='<%= PropsUtil.get("editor.wysiwyg.portal-impl.portlet.ddm.text_html.ftl") %>'
```
Comment (jbalsas) [inline modules/apps/journal/journal-web/src/main/resources/META-INF/resources/article/basic_info.jsp:96]: This is not our call to make. On top of that, this is not `ddm.text_html` so I don't think it's expected that it changes with that property

**Comment by jbalsas** [comment]
> To me the idea for ckeditor_classic is not to reproduce everything present in current ckeditor but to add things that are needed in the best way even if they don't follow the same API and we need to change the usages.

I agree with @carloslancha here. I'd just close this as `Won't Fix` for now and we'll analyze how to integrate it properly in the next release cycle.

File: modules/apps/journal/journal-web/src/main/resources/META-INF/resources/article/basic_info.jsp (hunk 1 above)
Comment (julien) [reply modules/apps/journal/journal-web/src/main/resources/META-INF/resources/article/basic_info.jsp:96]: @jbalsas thanks, I understand **we** shouldn't change it, but it did change something. I'm going to take a deeper look

**Comment by julien** [comment]
@jbalsas ok perfect, thanks for making that clear

File: modules/apps/journal/journal-web/src/main/resources/META-INF/resources/article/basic_info.jsp (hunk 1 above)
Comment (markocikos) [reply modules/apps/journal/journal-web/src/main/resources/META-INF/resources/article/basic_info.jsp:96]: It is up to the echo team if they want basic info option to have a customizable editor or if it's hardcoded to `ckeditor`. If they do want a customizable editor, we would would need to implement beckend for it, for example add new property
```
editor.wysiwyg.portal-web.docroot.html.portlet.journal.basic_info.jsp=ckeditor
```
etc.

**Review by markocikos — COMMENTED** [review]
We paused development on `ckeditor_classic` with only features needed for Message Boards. It needs more development, through implementation on other pages. The component is not mature enough to be expected to be safely added anywhere through `portal-ext.properties`. Looking at [portal.properties help comment](https://github.com/liferay/liferay-portal/blob/master/portal-impl/src/portal.properties#L5345-L5348), it is not listed as an option, which is correct. 

I wouldn't consider this a bug. Maybe we need to inform QA not to test this case, for now. I am not sure if we have an epic queued up to continue development.

**Comment by julien** [comment]
@markocikos thanks, I think we all agree, and I also think that this "bug" was created for us to investigate the possible places `ckeditor_classic` wasn't working as expected. 

As for the change in journal I made before, I understood that this is team decision and not ours.

I'm going to push more changes, because I did manage to fix the errors but I want to discuss them.

**Comment by julien** [comment]
Let's imagine that someone does decide to use `ckeditor_classic` by adding

```properties
editor.wysiwyg.portal-impl.portlet.ddm.text_html.ftl=ckeditor_classic
```
to his `portal-ext.properties` because he knows it's a thing.
*(I know @markocikos will say they shouldn't but I'm just trying to illustrate the
issue.)*

What's going to happen, is that we're both going to have `ckeditor_classic.jsp` and
`ckeditor.jsp` in the same "page", when that's the case what I've observed is a strange 
behaviour:

The `CKEDITOR.instances` object which is used a "registry" for editors is not
populated the same way it is when we only have `ckeditor.jsp` in the page.

When only having `ckeditor.jsp` in the page, `CKEDITOR.instances` will have
all the editor instances on the page, for example:

```js
CKEDITOR.instances = {
	_com_liferay_content_editor,  // An editor instance used for the content
	_com_liferay_description_editor,  // Another editor instance in the page
}
```

When both `ckeditor_classic.jsp` and `ckeditor.jsp` are present, this is what
I see 

```js
CKEDITOR.instances = {
	_com_liferay_content_editor,  // An editor instance used for the content
}
```

This is just to illustrate the problem.

What happens next, is that the
[`ckEditor`](https://github.com/liferay/liferay-portal/blob/2d2e8ae74e9e7b33a07328277da64f54db576a04/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/ckeditor.jsp#L488) variable that is used in
`ckeditor.jsp` (and other places in our `frontend-editor-ckeditor-web` module) is
`undefined` because `CKEDITOR.instances` does not contain it.

Since that `ckEditor` variable is `undefined` and used in
[`anchor_dialog_show.js`](https://github.com/liferay/liferay-portal/blob/2d2e8ae74e9e7b33a07328277da64f54db576a04/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/extension/anchor_dialog_show.js#L15), 
[`creole_dialog_definition.js`](https://github.com/liferay/liferay-portal/blob/2d2e8ae74e9e7b33a07328277da64f54db576a04/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/extension/creole_dialog_definition.js#L18)
and [`dialog_definition.js`](https://github.com/liferay/liferay-portal/blob/2d2e8ae74e9e7b33a07328277da64f54db576a04/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/extension/dialog_definition.js#L16)
all those can fail and throw an error. 


`CKEDITOR.instances` is not modified when you call `CKEDITOR.replace`; at
least not "straight away" in the `replace` function. 

As far as I've seen it gets modified by the
[`CKEDITOR.add`](https://github.com/ckeditor/ckeditor4/blob/aa1e0e7205cab2f4b6f917db1beaf0d438329114/core/ckeditor.js#L43), 
which gets called by [`Editor`](https://github.com/ckeditor/ckeditor4/blob/aa1e0e7205cab2f4b6f917db1beaf0d438329114/core/editor.js#L191) 
which is created by [`createInstance`](https://github.com/ckeditor/ckeditor4/blob/aa1e0e7205cab2f4b6f917db1beaf0d438329114/core/creators/themedui.js#L345)
which is in turn called by the [`replace`](https://github.com/ckeditor/ckeditor4/blob/major/core/creators/themedui.js#L39)

(For `CKEDITOR.inline` it's mostly almost the same).

What I did [do](https://github.com/liferay-frontend/liferay-portal/pull/390/commits/b24bf3ed06becded1feedc9afc03ed6bc992eb49#diff-da25dc049971ff4bf74778ef84eac3c0R482) (and I'm not saying it's the best way to solve it) 
is assign the value of `CKEDITOR.replace(...)` to the `ckEditor` variable, 
because they return the same thing, and this at least in this context (with
`ckeditor_classic.jsp` and `ckeditor.jsp` on the same page) fixes the problem, 
I also had to make a second change [here](https://github.com/liferay-frontend/liferay-portal/pull/390/commits/b24bf3ed06becded1feedc9afc03ed6bc992eb49#diff-da25dc049971ff4bf74778ef84eac3c0R589).

On top of that, `CKEDITOR.instances` is used in a lot of places and that's
obiously something we don't want to break.

As I said, this might not be the way we want to solve things, but I think it's
interesting enough to talk about it.

I still don't understand why I can't see two entries in `CKEDITOR.instances` so
I'm going to keep on digging.

**Comment by julien** [comment]
What's happening with `CKEDITOR.instances` must be DXP specific (I still haven't found where that's happening), because I've created a [codesandbox](https://codesandbox.io/s/sweet-mirzakhani-kq965?file=/public/index.html) example with two editors: one created directly with `CKEDITOR.replace` and the other in a React component and I can see the two instances.

**Comment by julien** [comment]
I managed to sport where this "bug" with `CKEDITOR.instances` is coming from:

[Here](https://github.com/liferay/liferay-portal/blob/1f1bccfda350d0fc3b94a904c6d54c8c32c6b341/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/ClassicEditor.js#L119), we're overriding the created editor's name with the `name` prop that was passed to the React component: 
So the "flow" goes something like this (bear with me because this is going to be hard to understand without debugging the code)

1. Render a `ClassicEditor` component that receives a `name` prop with a value of `com_liferay_some_editor`
2. `ClassicEditor` renders `Editor` which in turn renders `CKEditor` (and doesn't do anything with that `name` prop)
3. CKEDITOR (not the component, the library) [generates](https://github.com/ckeditor/ckeditor4/blob/aa1e0e7205cab2f4b6f917db1beaf0d438329114/core/editor.js#L96) the editor's name because the element created by the `CKEditor` component doesn't have a [`id` or `name`](https://github.com/ckeditor/ckeditor4/blob/aa1e0e7205cab2f4b6f917db1beaf0d438329114/core/editor.js#L68) attribute (*more about that later)
4. In the `CKEDITOR.on('instanceCreated')` callback we override "all/any" editor's name; setting it's value to whatever was passed in the `name` property. 

For me the first issue is that the `CKEditor` react component doesn't set an `id` or `name` to the element it creates. 
You can see this [here](https://github.com/ckeditor/ckeditor4-react/blob/d2686cc3e2b09b849ddf50b9c121dc8563915ffa/src/ckeditor.jsx#L51) but I also created another [codesandbox](https://codesandbox.io/s/focused-dream-pp6xo) to demonstrate this: just inspect `CKEDITOR.instances` in the codesanbox console and you'll notice it will always contain "editor1" instead of the name you're trying to pass to the editor (and I think I tried in any possible way - even though we know why this isn't happening). I'm probably going to open an issue on the [ckeditor4-react](https://github.com/ckeditor/ckeditor4/issues) repository if you all agree.

The second thing I don't understand right now (not saying it's wrong), is why we're listening to `CKEDITOR.on('instanceCreated')` instead of listenening to `editor.on('instanceCreated)` in our `ClassicEditor` component: because that `CKEDITOR.on('instanceCreated')` callback will be applied for all CKEDITOR editors not only this `ClassicEditor` as you can see in this screenshot, I still have to figure how to fix that,

**Comment by julien** [comment]
So `editor.on('instanceCreated')` doesn't exist to start with :facepalm: 
But we can use a `onInstanceReady` and `onDrop` prop on the component, or so it [seems](https://github.com/ckeditor/ckeditor4-react/blob/d2686cc3e2b09b849ddf50b9c121dc8563915ffa/src/ckeditor.jsx#L103) so I'm going to try that

**Comment by julien** [comment]
I'm [now](https://github.com/liferay-frontend/liferay-portal/pull/390/commits/ee0ca6672031f92f1a5325c35e8131d9b22a15b2) passing `onDrop` and `onInstanceReady` props to the component, everything works as expected.

When you can tell me what you think about creating and maybe submitting a patch to the CKEditor component to fix the fact that no name is being passed to the element it creates.

I'll probably squash some commits and drop some changes and force push again, because the only changes needed are for `ClassicEditor.js` and `ckeditor_classic.jsp`

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:test:relevant

**Comment by dsanz** [comment]
When I reported this issue, I did not considered much the current status of the development, meaning that I did not check whether it's the proper time to report/solve it or not.

Considering what @carloslancha , @markocikos  and @jbalsas said, I would say this editor is in a "beta" state, so to speak. I'd lean towards keeping what @julien discovered about coexistence of different ckeditors, as this is something we'll have to deal with sooner or later and will be helpful in the future. Additionally, I'll follow up on the plans we have to properly define the features we want ckeditor_classic to support, and it's properly integrated wherever needed. And consequently, to make it clear that it should not be used neither supported for our customers in this dxp release.

**Comment by julien** [comment]
@dsanz thanks for the feedback, I don't think it's a problem  having reported the issue even though `ckeditor_classic` is in beta.
This has been interesting because I was able to spot  a few things that should be changed, and as the saying goes: better safe than sorry.

I should probably update the description because right now it contains "stale" information, I was confused several times while trying to fix the issue, but the real "problem" is what I described [here](https://github.com/liferay-frontend/liferay-portal/pull/390#issuecomment-694150756) and which is:
 
- Don't override each CKEDITOR instance's name
- Avoid `CKEDITOR.on('instanceCreated', ...)` and prefer passing props that the component supports
- Send a fix (or at least open an issue) to the [ckeditor4-react](https://github.com/ckeditor/ckeditor4-react) repository

**Comment by julien** [comment]
ci:test:sf

**Comment by jbalsas** [comment]
These changes look reasonable to me ;)

**Comment by julien** [comment]
ci:test:relevant

**Comment by julien** [comment]
@john-co I ran both "failed" poshi tests in local to try understanding what was going on but I wasn't able to see anything. I also ran them in the `master` branch and in my case they were failing too - so I was wondering if you could have a look when you have time and tell me if these errors are related to my changes.

Thanks!

**Comment by julien** [comment]
ci:test:bundle

**Comment by wincent** [comment]
> @john-co I ran both "failed" poshi tests in local to try understanding what was going on but I wasn't able to see anything. I also ran them in the `master` branch and in my case they were failing too - so I was wondering if you could have a look when you have time and tell me if these errors are related to my changes.

Don't forget to mention @PablitoBonito as well on all frontend PRs that need help from QA (he's in local time zone, so you may get an answer sooner).

**Comment by PablitoBonito** [comment]
ci:test:bundle

**Comment by john-co** [comment]
Hi @julien, it looks like nothing happens after clicking on the image card in Item Selector modal:
https://issues.liferay.com/browse/LPS-121231

**Comment by julien** [comment]
Hi @John-Co, thanks that was helpful.

I figured out what was going on, but I don't have any idea of how to fix it; at least no "correct way".

This is also happening because of what I described [here](https://github.com/liferay-frontend/liferay-portal/pull/390#issuecomment-694150756), you can also see the [issue](https://github.com/ckeditor/ckeditor4-react/issues/125) I opened on the `ckeditor4-react` repository, but the "TL;DR" version is: 

*When using the CKEditor React (from `ckeditor4-react`) the editor's name is generated* (in this case it will always have the value `editor1` - and yes that's a problem)

So here's what I saw:

- When we click on the **link** button, CKEditor's "link" dialog is opened (I know that's obvious).

- When we click on the **browse server** button in that link dialog, [this](https://github.com/liferay/liferay-portal/blob/2a7b65296fedbccdd7899ec4c94919969882c6b2/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/itemselector/plugin.js#L364-L376) function is invoked:

  You'll see that the `url` argument passed to `openSelectionModal` function has the value of `editor.config.filebrowserBrowseUrl`, and in this case, that URL is defined [here](https://github.com/liferay/liferay-portal/blob/bff7584f2f9a4ab0ed7c1c3e51724ec09c9fd04b/modules/apps/item-selector/item-selector-editor-configuration/src/main/java/com/liferay/item/selector/editor/configuration/internal/DocumentsAndMediaURLEditorConfigContributor.java#L68-L70), 

- The next thing that happens is that the modal is created, opens/loads the correct URL and displays our images

- When we click on any of the cards, [this](https://github.com/liferay/liferay-portal/blob/b7c75041856c95fb220322d84de68e3c16875dfc/modules/apps/item-selector/item-selector-taglib/src/main/resources/META-INF/resources/repository_entry_browser/js/ItemSelectorRepositoryEntryBrowser.es.js#L363-L370) function is invoked 

- Then [this](https://github.com/liferay/liferay-portal/blob/652299bdd8db4bb6c26442d8c52fe5a9aa7f46b4/modules/apps/item-selector/item-selector-taglib/src/main/resources/META-INF/resources/repository_entry_browser/page.jsp#L641-L644) function/callback is invoked because the `itemSelector` is listening to the `selectedItem` event, and in turn dispatches another event, and that's where it "breaks": because the name of that event ( `'<%= itemSelectedEventName %>' `) is actually composed with the "wrong" editor name

So the "fix", I applied was simply modifying the URL defined [here](https://github.com/liferay/liferay-portal/blob/bff7584f2f9a4ab0ed7c1c3e51724ec09c9fd04b/modules/apps/item-selector/item-selector-editor-configuration/src/main/java/com/liferay/item/selector/editor/configuration/internal/DocumentsAndMediaURLEditorConfigContributor.java#L68-L70), to be:

```java
PortletURL itemSelectorURL = _itemSelector.getItemSelectorURL(
    requestBackedPortletURLFactory, "editor1selectItem",
    fileItemSelectorCriterion, layoutItemSelectorCriterion);
```

:see_no_evil: Of course that's not what we want to do ... but with that change, it works. 

I would suggest putting this PR "on hold", and either send a PR for [the](https://github.com/ckeditor/ckeditor4-react/issues/125) issue I created, or wait for that issue to be fixed by the `ckeditor4-react` team because:

1) this is a breaking change (or at least it seems like one) 
2) who knows what other "EdtorConfigContributor"s could break with this kind of change.

If you think there's a way to get the editor's name (no matter how it's has been generated) when we generate the URL, I'd be happy to know it.

/cc @jbalsas

**Comment by jbalsas** [comment]
Let's just park this for now.

**Comment by jbalsas** [comment]
Same as with others, I'm going to close this one just so it's more obvious what amount of PRs here need attention since this is starting to get out of hand.

**Pull request closed without merging**